include/oneapi/dpl/pstl/iterator_impl.h

```diff
@@ -401,7 +401,10 @@ class transform_iterator
 {
   private:
     _Iter __my_it_;
-    const _UnaryFunc __my_unary_func_;
+    _UnaryFunc __my_unary_func_;
+
+    static_assert(std::is_invocable_v<const _UnaryFunc, typename std::iterator_traits<_Iter>::reference>,
```

Comment:
Do we support function as the ref?

In other words: Does `transform_iterator<whatever_it, whatever_callable&>` work? If yes, my understanding is that the `static_assert` is broken.

Reply:
This is an interesting question... 

Before the addition of the `std::move` to the ctors, it seems that this "works" in practice to use `whatever_callable&`, at least on the host.  Further, this circumvents the `const` on the `operator*` call, because if we just have a ref as the type, the ref wont change even if its data inside does change, so `const` is no longer required for its `operator()` by our code.  

If anyone was previously using a ref type as the functor, adding `std::move` to the constructors seems to be a breaking change. 

Looking at the documentation and spec, there is nothing that specifically disallows using a `whatever_callable&`, but we do talk about it as a "unary functor/function".  It seems to be at least a stretch, and probably not something we want to support.  I would expect it is problematic for use on the device. 

Whether or not we add `std::move`, I think its best to add `std::decay_t<_UnaryFunc>` to the `static_assert` to make sure that our intended restriction is enforced.  

Note: 
For someone to be using a ref type they would have to do so explicitly by defining the `transform_iterator` template argument as a ref.  Sending a ref in to the constructor with deduction or to `make_transform_iterator` results in the decayed type being used.